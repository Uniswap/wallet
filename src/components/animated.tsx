import React from 'react'
import Animated from 'react-native-reanimated'

// TODO: [MOB-3865] find a way keep WrappedComponent props in return type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function withAnimated(WrappedComponent: React.ComponentType<any>): any {
  const displayName = WrappedComponent.displayName || WrappedComponent.name || 'Component'

  class WithAnimated extends React.Component {
    static displayName = `WithAnimated(${displayName})`

    render(): React.ReactNode {
      return <WrappedComponent {...this.props} />
    }
  }

  return Animated.createAnimatedComponent(WithAnimated)
}
