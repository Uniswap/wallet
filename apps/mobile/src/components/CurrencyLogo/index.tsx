import React from 'react'
import { TokenLogo } from 'src/components/CurrencyLogo/TokenLogo'
import { CurrencyInfo } from 'src/features/dataApi/types'
import { iconSizes } from 'src/styles/sizing'

interface CurrencyLogoProps {
  currencyInfo: NullUndefined<CurrencyInfo>
  size?: number
  hideNetworkLogo?: boolean
}

export const STATUS_RATIO = 2 / 5

export function CurrencyLogo({
  currencyInfo,
  size = iconSizes.icon40,
  hideNetworkLogo,
}: CurrencyLogoProps): JSX.Element | null {
  if (!currencyInfo) return null

  const { currency, logoUrl } = currencyInfo
  const { chainId, symbol } = currency

  return (
    <TokenLogo
      chainId={chainId}
      hideNetworkLogo={hideNetworkLogo}
      size={size}
      symbol={symbol}
      url={logoUrl}
    />
  )
}
