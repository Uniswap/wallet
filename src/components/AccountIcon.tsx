import React from 'react'
import { StyleSheet, useColorScheme } from 'react-native'
import Svg, { Defs, RadialGradient as RadialGradientSVG, Rect, Stop } from 'react-native-svg'
import { useAppTheme } from 'src/app/hooks'
import Eye from 'src/assets/icons/eye.svg'
import { RemoteImage } from 'src/components/images/RemoteImage'
import { Box } from 'src/components/layout'
import { Unicon } from 'src/components/unicons/Unicon'
import { useUniconColors } from 'src/components/unicons/utils'
import { theme as FixedTheme } from 'src/styles/theme'

interface Props {
  size: number
  showViewOnlyBadge?: boolean
  address: string
  avatarUri?: string | null
  showBackground?: boolean // Display images with solid background.
}

const INSET_PADDING = FixedTheme.spacing.spacing16

export function AccountIcon({
  size,
  showViewOnlyBadge,
  address,
  avatarUri,
  showBackground,
}: Props): JSX.Element {
  const theme = useAppTheme()

  // If background, add padding and center Unicons. Leave ENS avatars as is.
  const shouldShowUniconInsetPadding = !avatarUri && showBackground

  // If Unicon and background, reduce size to account for added padding.
  const adjustedIconSize = shouldShowUniconInsetPadding ? size - INSET_PADDING * 2 : size

  // Color for gradient background.
  const { gradientStart: uniconColor } = useUniconColors(address)

  const iconPadding = FixedTheme.spacing.spacing2
  const iconEyeContainerSize = size * 0.45
  const iconEyeSize = iconEyeContainerSize - iconPadding

  const isDarkMode = useColorScheme() === 'dark'

  const defaultImage = (
    <>
      <Unicon address={address} size={adjustedIconSize} />
      {showBackground ? <UniconGradient color={uniconColor} size={size} /> : null}
    </>
  )

  return (
    <Box
      backgroundColor={showBackground ? 'background0' : 'none'}
      borderColor={showBackground ? 'background0' : 'none'}
      borderRadius="roundedFull"
      borderWidth={showBackground ? 2 : 0}
      position="relative"
      style={{
        padding: shouldShowUniconInsetPadding ? INSET_PADDING : FixedTheme.spacing.none,
      }}>
      {avatarUri ? (
        <RemoteImage
          borderRadius={adjustedIconSize}
          fallback={defaultImage}
          height={adjustedIconSize}
          uri={avatarUri}
          width={adjustedIconSize}
        />
      ) : (
        defaultImage
      )}
      {showViewOnlyBadge && (
        <Box
          alignContent="center"
          alignItems="center"
          backgroundColor={isDarkMode ? 'background3' : 'background1'}
          borderRadius="roundedFull"
          bottom={0}
          height={iconEyeContainerSize}
          justifyContent="center"
          p="spacing4"
          position="absolute"
          right={0}
          shadowColor="black"
          shadowOffset={{ width: 0, height: 0 }}
          shadowOpacity={0.2}
          shadowRadius={10}
          width={iconEyeContainerSize}>
          <Eye color={theme.colors.textTertiary} width={iconEyeSize} />
        </Box>
      )}
    </Box>
  )
}

// Circle shaped gradient that follows Unicon colors.
const UniconGradient = ({ color, size }: { color: string; size: number }): JSX.Element => {
  return (
    <Svg height={size} style={UniconGradientStyles.svg} width={size}>
      <Defs>
        <RadialGradientSVG cy="-0.1" id="background" rx="0.8" ry="1.1">
          <Stop offset="0" stopColor={color} stopOpacity="0.6" />
          <Stop offset="1" stopColor={color} stopOpacity="0" />
        </RadialGradientSVG>
      </Defs>
      <Rect
        fill="url(#background)"
        height="100%"
        opacity={0.6}
        rx={size}
        width="100%"
        x="0"
        y="0"
      />
    </Svg>
  )
}

const UniconGradientStyles = StyleSheet.create({
  svg: {
    position: 'absolute',
  },
})
