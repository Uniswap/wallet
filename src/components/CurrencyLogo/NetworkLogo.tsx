import { SpacingProps, SpacingShorthandProps } from '@shopify/restyle'
import React from 'react'
import { Image } from 'react-native'
import { style } from 'src/components/CurrencyLogo/styles'
import { Box } from 'src/components/layout/Box'
import { ChainId, CHAIN_INFO } from 'src/constants/chains'
import { iconSizes } from 'src/styles/sizing'
import { Theme } from 'src/styles/theme'
import { useNetworkColors } from 'src/utils/colors'

type NetworkLogoProps = {
  chainId: ChainId
  borderWidth?: number
  size?: number
} & SpacingProps<Theme> &
  SpacingShorthandProps<Theme>

export function NetworkLogo({
  chainId,
  borderWidth = 0,
  size = iconSizes.icon20,
  ...rest
}: NetworkLogoProps): JSX.Element {
  const { logo } = CHAIN_INFO[chainId]
  const { foreground } = useNetworkColors(chainId)

  return (
    <Box
      borderRadius="roundedFull"
      borderWidth={borderWidth}
      style={{ borderColor: foreground }}
      {...rest}>
      {logo && (
        <Image
          source={logo}
          style={[
            style.image,
            {
              width: size,
              height: size,
              borderRadius: size / 2,
            },
          ]}
        />
      )}
    </Box>
  )
}
