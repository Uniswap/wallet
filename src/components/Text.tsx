import { createText, useResponsiveProp } from '@shopify/restyle'
import React, { ComponentProps, PropsWithChildren } from 'react'
import { useWindowDimensions } from 'react-native'
import Animated from 'react-native-reanimated'
import { Box } from 'src/components/layout'
import { Shimmer } from 'src/components/loading/Shimmer'
import { HiddenFromScreenReaders } from 'src/components/text/HiddenFromScreenReaders'
import { textVariants } from 'src/styles/font'
import { Theme } from 'src/styles/theme'

export const DEFAULT_FONT_SCALE = 1

export type TextProps = ComponentProps<typeof ThemedText> & {
  maxFontSizeMultiplier?: number
  animated?: boolean
  allowFontScaling?: boolean
  loading?: boolean | 'no-shimmer'
  loadingPlaceholderText?: string
}

// Use this text component throughout the app instead of
// Default RN Text for theme support
const ThemedText = createText<Theme>()
const ThemedAnimatedText = createText<Theme>(Animated.Text)

const TextPlaceholder = ({ children }: PropsWithChildren<unknown>): JSX.Element => {
  return (
    <Box alignItems="center" flexDirection="row">
      <Box alignItems="center" flexDirection="row" position="relative">
        <HiddenFromScreenReaders>{children}</HiddenFromScreenReaders>
        <Box
          bg="background3"
          borderRadius="rounded4"
          bottom="5%"
          left={0}
          position="absolute"
          right={0}
          top="5%"
        />
      </Box>
    </Box>
  )
}

const TextLoaderWrapper = ({
  children,
  loadingShimmer,
}: { loadingShimmer?: boolean } & PropsWithChildren<unknown>): JSX.Element => {
  const inner = <TextPlaceholder>{children}</TextPlaceholder>
  if (loadingShimmer) {
    return <Shimmer>{inner}</Shimmer>
  }

  return inner
}

/**
 * Use this component instead of the default React Native <Text> component anywhere text shows up throughout the app, so we can use the design system values for colors and sizes, and make sure all text looks and behaves the same way
 * @param loading Whether the text inside the component is still loading or not. Set this to true if whatever content goes inside the <Text> component is coming from a variable that might still be loading. This prop is optional and defaults to false. This prop can also be set to "no-shimmer" to enable a loading state without the shimmer effect.
 * @param loadingPlaceholderText - The text that the loader's size will be derived from. Pick something that's close to the same length as the final text is expected to be, e.g. if it's a ticker symbol, "XXX" might be a good placeholder text. This prop is optional and defaults to "$00.00".
 */
export const Text = ({
  animated,
  loading = false,
  maxFontSizeMultiplier,
  allowFontScaling,
  loadingPlaceholderText = '$00.00',
  ...rest
}: TextProps): JSX.Element => {
  const { fontScale } = useWindowDimensions()
  const variant = useResponsiveProp(rest.variant ?? 'bodySmall') as keyof typeof textVariants
  const enableFontScaling = allowFontScaling ?? fontScale > DEFAULT_FONT_SCALE
  const multiplier = maxFontSizeMultiplier ?? textVariants[variant].maxFontSizeMultiplier

  if (animated) {
    return (
      <ThemedAnimatedText
        allowFontScaling={enableFontScaling}
        maxFontSizeMultiplier={multiplier}
        {...rest}
      />
    )
  }

  if (loading) {
    return (
      <TextLoaderWrapper loadingShimmer={loading !== 'no-shimmer'}>
        <ThemedText
          allowFontScaling={enableFontScaling}
          color="none"
          maxFontSizeMultiplier={multiplier}
          opacity={0}
          {...rest}>
          {/* Important that `children` isn't used or rendered by <Text> when `loading` is true, because if the child of a <Text> component is a dynamic variable that might not be finished fetching yet, it'll result in an error until it's finished loading. We use `loadingPlaceholderText` to set the size of the loading element instead. */}
          {loadingPlaceholderText}
        </ThemedText>
      </TextLoaderWrapper>
    )
  }

  return (
    <ThemedText allowFontScaling={enableFontScaling} maxFontSizeMultiplier={multiplier} {...rest} />
  )
}
