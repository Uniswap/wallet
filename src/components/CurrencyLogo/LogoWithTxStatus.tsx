import React from 'react'
import { StyleSheet } from 'react-native'
import { useAppTheme } from 'src/app/hooks'
import AlertTriangle from 'src/assets/icons/alert-triangle.svg'
import Approve from 'src/assets/icons/approve.svg'
import IncomingArrow from 'src/assets/icons/arrow-down-in-circle.svg'
import OutgoingArrow from 'src/assets/icons/arrow-up-in-circle.svg'
import UnknownStatus from 'src/assets/icons/question-in-circle.svg'
import SlashCircleIcon from 'src/assets/icons/slash-circle.svg'
import WalletConnectLogo from 'src/assets/icons/walletconnect.svg'
import MoonpayLogo from 'src/assets/logos/moonpay.svg'
import { CurrencyLogo } from 'src/components/CurrencyLogo'
import { NetworkLogo } from 'src/components/CurrencyLogo/NetworkLogo'
import { ImageUri } from 'src/components/images/ImageUri'
import { NFTViewer } from 'src/components/images/NFTViewer'
import { RemoteImage } from 'src/components/images/RemoteImage'
import { Box } from 'src/components/layout/Box'
import { DappIconPlaceholder } from 'src/components/WalletConnect/DappHeaderIcon'
import { ChainId } from 'src/constants/chains'
import { AssetType } from 'src/entities/assets'
import { CurrencyInfo } from 'src/features/dataApi/types'
import { NFTTradeType, TransactionStatus, TransactionType } from 'src/features/transactions/types'
import { WalletConnectEvent } from 'src/features/walletConnect/saga'
import { logger } from 'src/utils/logger'

interface LogoWithTxStatusProps {
  assetType: AssetType
  txType: TransactionType
  txStatus: TransactionStatus
  size: number
}

interface DappLogoWithTxStatusProps {
  event: WalletConnectEvent
  size: number
  chainId: ChainId | null
  dappImageUrl: string | null
  dappName: string
}

interface SwapLogoOrLogoWithTxStatusProps {
  inputCurrencyInfo: NullUndefined<CurrencyInfo>
  outputCurrencyInfo: NullUndefined<CurrencyInfo>
  txStatus: TransactionStatus
  size: number
  showCancelIcon?: boolean
}

interface CurrencyStatusProps extends LogoWithTxStatusProps {
  assetType: AssetType.Currency
  currencyInfo?: CurrencyInfo | null
}

interface NFTStatusProps extends LogoWithTxStatusProps {
  assetType: AssetType.ERC721 | AssetType.ERC1155
  nftImageUrl?: string
  nftTradeType?: NFTTradeType
}

export function LogoWithTxStatus(props: CurrencyStatusProps | NFTStatusProps): JSX.Element {
  const { assetType, txType, txStatus, size } = props
  const theme = useAppTheme()

  const statusSize = size * (1 / 2)

  const logo =
    txType === TransactionType.FiatPurchase ? (
      <MoonpayLogo width={size} />
    ) : assetType === AssetType.Currency ? (
      <CurrencyLogo hideNetworkLogo currencyInfo={props.currencyInfo} size={size} />
    ) : (
      <Box
        alignItems="center"
        backgroundColor="background2"
        borderRadius="rounded4"
        height={size}
        justifyContent="center"
        overflow="hidden"
        width={size}>
        {props.nftImageUrl && <NFTViewer uri={props.nftImageUrl} />}
      </Box>
    )

  const fill = theme.colors.background0
  const gray = theme.colors.textSecondary
  const green = theme.colors.accentSuccess
  const yellow = theme.colors.accentWarning

  const getTxStatusIcon = (): JSX.Element | null => {
    if (txStatus === TransactionStatus.Failed) {
      return <AlertTriangle color={yellow} fill={fill} height={statusSize} width={statusSize} />
    }
    if (txStatus === TransactionStatus.Cancelled) {
      return (
        <SlashCircleIcon
          color={theme.colors.textSecondary}
          fill={theme.colors.background0}
          fillOpacity={1}
          height={statusSize}
          width={statusSize}
        />
      )
    }
    switch (txType) {
      case TransactionType.Approve:
        return <Approve color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.NFTApprove:
        return <Approve color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.Send:
        return <OutgoingArrow color={green} fill={fill} height={statusSize} width={statusSize} />
      // Fiat purchases use the same icon as receive
      case TransactionType.FiatPurchase:
      case TransactionType.Receive:
        return <IncomingArrow color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.NFTMint:
        return <IncomingArrow color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.NFTTrade:
        if (assetType === AssetType.ERC721 && props.nftTradeType === NFTTradeType.SELL) {
          return <OutgoingArrow color={green} fill={fill} height={statusSize} width={statusSize} />
        }
        return <IncomingArrow color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.Unknown:
        return <UnknownStatus color={gray} fill={fill} height={statusSize} width={statusSize} />
    }
    logger.debug(
      'statusIcon',
      'GenerateStatusIcon',
      'Could not find icon for transaction type:',
      txType
    )
    return null
  }
  const statusIcon = getTxStatusIcon()
  return (
    <Box alignItems="center" height={size} justifyContent="center" width={size}>
      {logo}
      <Box bottom={-4} position="absolute" right={-4}>
        {statusIcon}
      </Box>
    </Box>
  )
}

export function SwapLogoOrLogoWithTxStatus({
  size,
  inputCurrencyInfo,
  outputCurrencyInfo,
  txStatus,
  showCancelIcon,
}: SwapLogoOrLogoWithTxStatusProps): JSX.Element {
  if (
    txStatus === TransactionStatus.Failed ||
    (showCancelIcon && txStatus === TransactionStatus.Cancelled)
  ) {
    return (
      <LogoWithTxStatus
        assetType={AssetType.Currency}
        currencyInfo={inputCurrencyInfo}
        size={size}
        txStatus={txStatus}
        txType={TransactionType.Swap}
      />
    )
  }

  return (
    <Box height={size} width={size}>
      <Box left={0} position="absolute" testID="swap-success-toast" top={0}>
        <CurrencyLogo hideNetworkLogo currencyInfo={inputCurrencyInfo} size={size * (2 / 3)} />
      </Box>
      <Box bottom={0} position="absolute" right={0}>
        <CurrencyLogo hideNetworkLogo currencyInfo={outputCurrencyInfo} size={size * (2 / 3)} />
      </Box>
    </Box>
  )
}

export function DappLogoWithTxStatus({
  dappName,
  dappImageUrl,
  event,
  size,
  chainId,
}: DappLogoWithTxStatusProps): JSX.Element | null {
  const theme = useAppTheme()
  const green = theme.colors.accentSuccess
  const yellow = theme.colors.accentWarning
  const fill = theme.colors.background0

  const dappImageSize = size
  const statusSize = dappImageSize * (1 / 2)
  const totalSize = dappImageSize + statusSize * (1 / 4)

  const getStatusIcon = (): JSX.Element | undefined => {
    switch (event) {
      case WalletConnectEvent.NetworkChanged:
        return chainId ? <NetworkLogo chainId={chainId} size={statusSize} /> : undefined
      case WalletConnectEvent.TransactionConfirmed:
        return <Approve color={green} fill={fill} height={statusSize} width={statusSize} />
      case WalletConnectEvent.TransactionFailed:
        return <AlertTriangle color={yellow} fill={fill} height={statusSize} width={statusSize} />
    }
  }

  const statusIcon = getStatusIcon()

  const fallback = (
    <Box height={dappImageSize}>
      <DappIconPlaceholder iconSize={dappImageSize} name={dappName} />
    </Box>
  )

  const style = StyleSheet.create({
    icon: {
      borderRadius: theme.borderRadii.rounded4,
      height: dappImageSize,
      width: dappImageSize,
    },
    loaderContainer: {
      borderRadius: theme.borderRadii.roundedFull,
      overflow: 'hidden',
    },
  })

  const dappImage = dappImageUrl ? (
    <ImageUri
      fallback={fallback}
      imageStyle={style.icon}
      loadingContainerStyle={{ ...style.icon, ...style.loaderContainer }}
      uri={dappImageUrl}
    />
  ) : (
    fallback
  )

  return statusIcon ? (
    <Box height={totalSize} width={totalSize}>
      <Box left={0} position="absolute" top={0}>
        {dappImage}
      </Box>
      <Box bottom={0} position="absolute" right={0}>
        {statusIcon}
      </Box>
    </Box>
  ) : (
    dappImage
  )
}

/** For displaying Dapp logo with generic WC bade icon */
export function DappLogoWithWCBadge({
  dappImageUrl,
  dappName,
  size,
}: {
  dappImageUrl: string | null
  dappName: string
  size: number
}): JSX.Element {
  const theme = useAppTheme()
  const dappImageSize = size
  const statusSize = dappImageSize * (1 / 2)
  const totalSize = dappImageSize + statusSize * (1 / 4)
  const dappImage = dappImageUrl ? (
    <RemoteImage
      borderRadius={theme.borderRadii.rounded4}
      height={dappImageSize}
      uri={dappImageUrl}
      width={dappImageSize}
    />
  ) : (
    <DappIconPlaceholder iconSize={dappImageSize} name={dappName} />
  )

  return (
    <Box height={totalSize} width={totalSize}>
      <Box left={4} position="absolute" top={0}>
        {dappImage}
      </Box>
      <Box
        backgroundColor="background1"
        borderRadius="rounded4"
        bottom={0}
        position="absolute"
        right={0}>
        <WalletConnectLogo fill="red" height={statusSize} width={statusSize} />
      </Box>
    </Box>
  )
}
