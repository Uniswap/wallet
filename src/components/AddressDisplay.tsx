import { default as React, PropsWithChildren, useMemo } from 'react'
import { FlexAlignType } from 'react-native'
import { useAppDispatch, useAppTheme } from 'src/app/hooks'
import CopyIcon from 'src/assets/icons/copy-sheets.svg'
import { AccountIcon } from 'src/components/AccountIcon'
import { TouchableArea } from 'src/components/buttons/TouchableArea'
import { Box, Flex } from 'src/components/layout'
import { Text } from 'src/components/Text'
import { useENSAvatar } from 'src/features/ens/api'
import { pushNotification } from 'src/features/notifications/notificationSlice'
import { AppNotificationType } from 'src/features/notifications/types'
import { ElementName } from 'src/features/telemetry/constants'
import { useDisplayName } from 'src/features/wallet/hooks'
import { Theme } from 'src/styles/theme'
import { sanitizeAddressText, shortenAddress } from 'src/utils/addresses'
import { setClipboard } from 'src/utils/clipboard'

type AddressDisplayProps = {
  address: string
  hideAddressInSubtitle?: boolean
  size?: number
  variant?: keyof Theme['textVariants']
  captionVariant?: keyof Theme['textVariants']
  direction?: 'row' | 'column'
  showCopy?: boolean
  showCopyWrapperButton?: boolean
  showAccountIcon?: boolean
  contentAlign?: FlexAlignType
  showIconBackground?: boolean
  textAlign?: FlexAlignType
  horizontalGap?: keyof Theme['spacing']
}

type CopyButtonWrapperProps = {
  onPress?: () => void
  backgroundColor?: string
}

function CopyButtonWrapper({
  children,
  onPress,
}: PropsWithChildren<CopyButtonWrapperProps>): JSX.Element {
  if (onPress)
    return (
      <TouchableArea
        hapticFeedback
        name={ElementName.Copy}
        testID={ElementName.Copy}
        onPress={onPress}>
        {children}
      </TouchableArea>
    )

  return <>{children}</>
}

/** Helper component to display identicon and formatted address */
export function AddressDisplay({
  address,
  size = 24,
  variant = 'bodyLarge',
  captionVariant = 'subheadSmall',
  hideAddressInSubtitle,
  direction = 'row',
  showCopy = false,
  showCopyWrapperButton = false,
  showAccountIcon = true,
  textAlign,
  contentAlign = 'center', // vertical aligment of all items
  showIconBackground,
  horizontalGap = 'spacing12',
}: AddressDisplayProps): JSX.Element {
  const dispatch = useAppDispatch()
  const theme = useAppTheme()
  const displayName = useDisplayName(address)
  const { data: avatar } = useENSAvatar(address)

  const showAddressAsSubtitle = !hideAddressInSubtitle && displayName?.type !== 'address'

  const onPressCopyAddress = (): void => {
    if (!address) return
    dispatch(pushNotification({ type: AppNotificationType.Copied }))
    setClipboard(address)
  }

  // Extract sizes so copy icon can match font variants
  const mainSize = theme.textVariants[variant].fontSize
  const captionSize = theme.textVariants[captionVariant].fontSize
  const itemAlignment =
    textAlign || (!showAccountIcon || direction === 'column' ? 'center' : 'flex-start')

  const icon = useMemo(() => {
    return (
      <AccountIcon
        address={address}
        avatarUri={avatar}
        showBackground={showIconBackground}
        size={size}
      />
    )
  }, [address, avatar, showIconBackground, size])

  return (
    <Flex alignItems={contentAlign} flexDirection={direction} gap={horizontalGap}>
      {showAccountIcon && icon}
      <Box alignItems={itemAlignment} flexShrink={1}>
        <CopyButtonWrapper
          onPress={showCopy && !showAddressAsSubtitle ? onPressCopyAddress : undefined}>
          <Flex centered row gap="spacing12">
            <Text
              color="textPrimary"
              ellipsizeMode="tail"
              numberOfLines={1}
              testID={`address-display/name/${displayName?.name}`}
              variant={variant}>
              {displayName?.name}
            </Text>
            {showCopy && !showAddressAsSubtitle && (
              <CopyIcon color={theme.colors.textPrimary} height={mainSize} width={mainSize} />
            )}
          </Flex>
        </CopyButtonWrapper>
        {showAddressAsSubtitle && (
          <CopyButtonWrapper onPress={showCopy ? onPressCopyAddress : undefined}>
            <Flex
              centered
              row
              backgroundColor={showCopyWrapperButton ? 'backgroundOverlay' : 'none'}
              borderRadius="roundedFull"
              gap="spacing4"
              marginTop={showCopyWrapperButton ? 'spacing8' : 'none'}
              px={showCopyWrapperButton ? 'spacing8' : 'none'}
              py={showCopyWrapperButton ? 'spacing4' : 'none'}>
              <Text color="textSecondary" variant={captionVariant}>
                {sanitizeAddressText(shortenAddress(address))}
              </Text>
              {showCopy && (
                <CopyIcon
                  color={theme.colors.textSecondary}
                  height={captionSize}
                  width={captionSize}
                />
              )}
            </Flex>
          </CopyButtonWrapper>
        )}
      </Box>
    </Flex>
  )
}
