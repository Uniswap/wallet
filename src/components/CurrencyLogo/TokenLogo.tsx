import React from 'react'
import { Image } from 'react-native'
import { useAppTheme } from 'src/app/hooks'
import { NetworkLogo } from 'src/components/CurrencyLogo/NetworkLogo'
import { SHADOW_OFFSET, style, THIN_BORDER } from 'src/components/CurrencyLogo/styles'
import { Box } from 'src/components/layout/Box'
import { Text } from 'src/components/Text'
import { ChainId } from 'src/constants/chains'
import { iconSizes } from 'src/styles/sizing'

interface TokenLogoProps {
  url?: string | null
  symbol?: string
  chainId?: ChainId
  size?: number
  hideNetworkLogo?: boolean
}

export function TokenLogo({
  url,
  symbol,
  chainId,
  size = iconSizes.icon40,
  hideNetworkLogo,
}: TokenLogoProps): JSX.Element {
  const theme = useAppTheme()
  const showNetworkLogo = !hideNetworkLogo && chainId && chainId !== ChainId.Mainnet

  return (
    <Box alignItems="center" height={size} justifyContent="center" width={size}>
      {url ? (
        <Image
          source={{ uri: url }}
          style={[
            style.image,
            {
              backgroundColor: theme.colors.backgroundOutline,
              width: size,
              height: size,
              borderRadius: size / 2,
              borderColor: theme.colors.backgroundOutline,
              borderWidth: THIN_BORDER,
            },
          ]}
        />
      ) : (
        <Box
          alignItems="center"
          bg="backgroundOutline"
          borderRadius="roundedFull"
          height={size}
          justifyContent="center"
          px="spacing8"
          width={size}>
          <Text adjustsFontSizeToFit color="textPrimary" numberOfLines={1} textAlign="center">
            {symbol?.slice(0, 3).toUpperCase()}
          </Text>
        </Box>
      )}
      {showNetworkLogo && (
        <Box
          bottom={0}
          position="absolute"
          right={0}
          shadowColor="black"
          shadowOffset={SHADOW_OFFSET}
          shadowOpacity={0.1}
          shadowRadius={2}>
          <NetworkLogo chainId={chainId} size={size * (2 / 5)} />
        </Box>
      )}
    </Box>
  )
}
