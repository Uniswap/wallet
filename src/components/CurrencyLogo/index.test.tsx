import React from 'react'
import { CurrencyLogo } from 'src/components/CurrencyLogo'
import { arbitrumDaiCurrencyInfo, uniCurrencyInfo } from 'src/test/fixtures'
import { render } from 'src/test/test-utils'

jest.mock('src/assets', () => 'ethereum-logo.png')

it('renders a currency logo without network logo', () => {
  const tree = render(<CurrencyLogo currencyInfo={uniCurrencyInfo} size={20} />)
  expect(tree).toMatchSnapshot()
})

it('renders a currency logo with network logo', () => {
  const tree = render(<CurrencyLogo currencyInfo={arbitrumDaiCurrencyInfo} size={20} />)
  expect(tree).toMatchSnapshot()
})
