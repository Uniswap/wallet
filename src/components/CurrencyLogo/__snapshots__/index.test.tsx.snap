// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a currency logo with network logo 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      },
    ]
  }
>
  <Image
    source={
      Object {
        "uri": "https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png",
      }
    }
    style={
      Array [
        Object {
          "resizeMode": "contain",
        },
        Object {
          "backgroundColor": "#D2D9EE",
          "borderColor": "#D2D9EE",
          "borderRadius": 10,
          "borderWidth": 0.5,
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "bottom": 0,
          "display": "flex",
          "position": "absolute",
          "right": 0,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 2,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "borderRadius": 999999,
            "borderWidth": 0,
            "display": "flex",
          },
          Object {
            "borderColor": "#515EDC",
          },
        ]
      }
    />
  </View>
</View>
`;

exports[`renders a currency logo without network logo 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      },
    ]
  }
>
  <Image
    source={
      Object {
        "uri": "https://raw.githubusercontent.com/Uniswap/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png",
      }
    }
    style={
      Array [
        Object {
          "resizeMode": "contain",
        },
        Object {
          "backgroundColor": "#D2D9EE",
          "borderColor": "#D2D9EE",
          "borderRadius": 10,
          "borderWidth": 0.5,
          "height": 20,
          "width": 20,
        },
      ]
    }
  />
</View>
`;
